### PR TITLE
Italian Language update

### PR DESCRIPTION
**Pull request by maurobennici**
The Italian extractor and parser are now updated.
All the tests are translated following the French examples.

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Definitions/Italian/NumbersDefinitions.tt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿<#@ template debug="true" hostspecific="true" language="C#" #>
+<#
+	this.DataFilename = @"Patterns\Italian\Italian-Numbers.yaml";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/Italian/NumbersDefinitions.tt]: I see that you have translated most json specs and that some Italian tests pass, but many fail.
I suggest you mark the current failing tests as "NotSupported": "javascript,dotnet" in the json files. This will allow the build to pass. 
As problems are fixed and new tests pass, you can remove "dotnet" from the NotSupported tag. As soon as each extractor/parser passes all its tests, we can merge little by little. No need to wait for the whole Italian recognizers to be ready at once.
This will make the review process much simpler, as well as allow others to contribute if they can.

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Definitions/Italian/NumbersDefinitions.tt]: I also suggest start with Choice and Numbers, then Units, and leaving the DateTime fixes for last as it's the most complex recognizer.
Once you get to it, them start there with the basic types.

File: .NET/Microsoft.Recognizers.Text.DateTime/DateTimeRecognizer.cs [hunk 2]
```diff
@@ -76,10 +79,10 @@ protected override void InitializeConfiguration()
                     new BaseMergedExtractor(new PortugueseMergedExtractorConfiguration(options))));
 
             RegisterModel<DateTimeModel>(
-                Culture.German,
-                (options) => new DateTimeModel(
-                    new BaseMergedParser(new GermanMergedParserConfiguration(options)),
-                    new BaseMergedExtractor(new GermanMergedExtractorConfiguration(options))));
+              Culture.Italian,
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/DateTimeRecognizer.cs]: In order to merge PRs for new languages, they should not be registered at the recognizer level until all their tests pass (or the majority, if there's already enough value-add).
Please remove the registration for now.

File: .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDatePeriodExtractorConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+        public static readonly Regex MonthWithYear =
+            new Regex(
+                DateTimeDefinitions.MonthWithYear,
+                RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+        public static readonly Regex OneWordPeriodRegex =
+            new Regex(
+               DateTimeDefinitions.OneWordPeriodRegex,
+                RegexOptions.IgnoreCase | RegexOptions.Singleline); // a cote de - 'next to', cette - 'this', dernier - 'last' (always after the noun, i.e annee dernier - 'last year'
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDatePeriodExtractorConfiguration.cs]: Nit: I'd prefer to move the code comment before the regex declaration.
It's easier to read that way.

File: .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDatePeriodExtractorConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+        public static readonly Regex DecadeWithCenturyRegex =
+            new Regex(DateTimeDefinitions.DecadeWithCenturyRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+        public static readonly Regex YearPeriodRegex =
+            new Regex(DateTimeDefinitions.YearPeriodRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+        public static readonly Regex RelativeDecadeRegex =
+            new Regex(DateTimeDefinitions.RelativeDecadeRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+        private static readonly Regex FromRegex = new Regex(DateTimeDefinitions.FromRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDatePeriodExtractorConfiguration.cs]: Nit: Please follow the same standard from the other regexes for these 3.

File: .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateTimePeriodExtractorConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        private static readonly Regex[] SimpleCases =
+        {
+            ItalianTimePeriodExtractorConfiguration.PureNumFromTo,
+            ItalianTimePeriodExtractorConfiguration.PureNumBetweenAnd,
+            ItalianTimePeriodExtractorConfiguration.SpecificTimeOfDayRegex
+        };
+
+        private static readonly Regex FromRegex = new Regex(DateTimeDefinitions.FromRegex2, RegexOptions.IgnoreCase | RegexOptions.Singleline);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateTimePeriodExtractorConfiguration.cs]: Same

File: .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateTimePeriodExtractorConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+        Regex IDateTimePeriodExtractorConfiguration.NumberCombinedWithUnit => TimeNumberCombinedWithUnit;
+
+        Regex IDateTimePeriodExtractorConfiguration.TimeUnitRegex => TimeTimeUnitRegex;
+
+        Regex IDateTimePeriodExtractorConfiguration.RelativeTimeUnitRegex => RelativeTimeUnitRegex;
+
+        Regex IDateTimePeriodExtractorConfiguration.RestOfDateTimeRegex => RestOfDateTimeRegex;
+
+        public Regex PastPrefixRegex => ItalianDatePeriodExtractorConfiguration.PastPrefixRegex; // Note: FR 'past' i.e 'dernier' is a suffix following after, however interface enforces 'prefix' nomenclature
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateTimePeriodExtractorConfiguration.cs]: Same for placement.
Also, the comments are in French. Can you translate them?

File: .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianHolidayExtractorConfiguration.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            new Regex(
+                DateTimeDefinitions.HolidayRegex2,
+                RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+        public static readonly Regex H3 =
+            new Regex(
+                DateTimeDefinitions.HolidayRegex3,
+                RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+        // added to include more options, "fete des meres" mothers day, etc
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianHolidayExtractorConfiguration.cs]: You can remove this commented out code if you're not going to use it.

File: .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianHolidayExtractorConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        //    new Regex(
+        //        DateTimeDefinitions.HolidayRegex4,
+        //        RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+        public static readonly Regex[] HolidayRegexList =
+        {
+            H1,
+            H2,
+            H3,
+            //H4
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianHolidayExtractorConfiguration.cs]: Same

File: Patterns/Italian/Italian-DateTime.yaml [hunk 9]
```diff
@@ -1,131 +1,127 @@
...
+  def: (settimana)(\s*)(?<number>\d\d|\d|0\d)
 WeekOfRegex: !simpleRegex
-  def: (the\s+)?(week)(\s+of)
+  def: (settimana)(\s*)(del)
 MonthOfRegex: !simpleRegex
-  def: (month)(\s*)(of)
+  def: (mese)(\s*)(di)
 MonthRegex: !simpleRegex
-  def: (?<month>April|Apr|August|Aug|December|Dec|February|Feb|January|Jan|July|Jul|June|Jun|March|Mar|May|November|Nov|October|Oct|September|Sept|Sep)
-AmDescRegex: !simpleRegex
```
Comment (tellarin) [inline Patterns/Italian/Italian-DateTime.yaml]: Why did you remove these other regexes? Please keep them if possible.

File: Patterns/Italian/Italian-DateTime.yaml [hunk 10]
```diff
@@ -1,131 +1,127 @@
...
   references: [ WeekDayRegex ]
 LastDateRegex: !nestedRegex
-  def: \b({PastPrefixRegex}(\s*week)?\s+{WeekDayRegex})|({WeekDayRegex}(\s+last\s*week))\b
-  references: [ PastPrefixRegex, WeekDayRegex ]
+  def: \b(({WeekDayRegex}(\s*(della)?\s*la\s*settimana\s+{PastSuffixRegex}))|({WeekDayRegex}(\s+{PastSuffixRegex})))\b
+  references: [ PastSuffixRegex, WeekDayRegex ]
 NextDateRegex: !nestedRegex
-  def: \b({NextPrefixRegex}(\s*week(\s*on)?)?\s+{WeekDayRegex})|((on\s+)?{WeekDayRegex}(\s+next\s*week))\b
-  references: [ NextPrefixRegex, WeekDayRegex ]
-SpecialDayRegex: !nestedRegex
```
Comment (tellarin) [inline Patterns/Italian/Italian-DateTime.yaml]: Same. Please don't remove regexes.

File: Patterns/Italian/Italian-DateTime.yaml [hunk 11]
```diff
@@ -134,336 +130,402 @@ DateExtractor5: !nestedRegex
...
 MonthEnd: !nestedRegex
-  def: '{MonthRegex}\s*(the)?\s*$'
+  def: '{MonthRegex}\s*(il)?\s*$'
   references: [ MonthRegex ]
+WeekDayEnd: !nestedRegex
+  def: '{WeekDayRegex}\s*,?\s*$'
+  references: [ WeekDayRegex ]
 RangeUnitRegex: !simpleRegex
-  def: \b(?<unit>years|year|months|month|weeks|week)\b
-OclockRegex: !simpleRegex
```
Comment (tellarin) [inline Patterns/Italian/Italian-DateTime.yaml]: Same. If unnecessary, you can set them to a placeholder.

Comment (tellarin) [reply Patterns/Italian/Italian-DateTime.yaml]: This will help keep the code between languages more similar. And avoid nullpointerexceptions in base code that might try to use the regexes.

Comment (tellarin) [reply Patterns/Italian/Italian-DateTime.yaml]: Feel free to add Italian-specific regexes as needed though.

File: Patterns/Italian/Italian-NumbersWithUnit.yaml [hunk 12]
```diff
@@ -330,7 +330,7 @@ CurrencySuffixList: !dictionary
 #Dollar
     United States dollar: dollaro americano|dollari americani|dollaro statunitense|dollari statunitensi|dollaro degli stati uniti|dollari delgi stati uniti|u.s. dollars|u.s. dollar|u s dollar|u s dollars|usd|us$|us dollar|us dollars|u.s dollar|u.s dollars
     East Caribbean dollar: dollaro dei caraibi orientali|dollari dei caraibi orientali|east Caribbean $|xcd
-    Australian dollar: dollaro australiano|dollari australiani|dollaro dell'australia|australian $|australian$|aud
+    Australian dollar: dollaro australiano|dollari australiani|dollaro dell australia|australian $|australian$|aud
```
Comment (tellarin) [inline Patterns/Italian/Italian-NumbersWithUnit.yaml]: The currency names should also be translated. See French and Spanish yamls.

File: Specs/DateTime/Italian/DateExtractor.json [hunk 13]
```diff
@@ -131,29 +120,7 @@
 }
 ,
 {
-  "Input": "I'll go back on 4.22",
```
Comment (tellarin) [inline Specs/DateTime/Italian/DateExtractor.json]: Try to keep as many of the original test cases if you can.
Only remove if they really don't make sense in Italian, and there is no other similar cases that could be used.

File: Specs/DateTime/Italian/DateExtractor.json [hunk 14]
```diff
@@ -285,889 +241,834 @@
 }
 ,
 {
-  "Input": "I'll go back Wed, 22 of Jan",
```
Comment (tellarin) [inline Specs/DateTime/Italian/DateExtractor.json]: It's preferable to mark test cases as NotSupported in dotnet and javascript temporarily, then to remove them.
This helps make sure the test coverage remains high.
As translations and fixes are in, we can remove the annotations. This way it's easier to track coverage.

File: Specs/DateTime/Italian/DatePeriodExtractor.json [hunk 15]
```diff
@@ -791,1437 +659,1261 @@
 }
 ,
 {
-  "Input": "I'll be out in Dec",
-  "NotSupportedByDesign": "python",
```
Comment (tellarin) [inline Specs/DateTime/Italian/DatePeriodExtractor.json]: Same

File: Specs/Number/Italian/NumberRangeModel.json [hunk 16]
```diff
@@ -6,464 +6,464 @@
...
+  "Input": "Questo numero è più grande di venti e minore o uguale a trentacinque.",
   "NotSupported": "javascript",
   "NotSupportedByDesign": "python",
   "Results": [
     {
-      "Text": "larger than twenty and less or equal than thirty five",
+      "Text": "più grande di venti e minore o uguale a trentacinque",
       "TypeName": "numberrange",
       "Resolution": {
-        "value": "(20,35]"
```
Comment (tellarin) [inline Specs/Number/Italian/NumberRangeModel.json]: The interval syntax should not be modified by language.
This is a math syntax, it can represent intervals like [-15,-1], for example.

Comment (tellarin) [reply Specs/Number/Italian/NumberRangeModel.json]: It would be ambiguous for the case of open intervals with negative bounds.

File: .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateExtractorConfiguration.cs [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+            new Regex(
+                DateTimeDefinitions.RelativeMonthRegex,
+                RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+        public static readonly Regex PrefixArticleRegex =
+            new Regex(
+                DateTimeDefinitions.PrefixArticleRegex,
+                RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+
```
[the remark this answers is not in the document]
Comment (maurobennici) [reply .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateExtractorConfiguration.cs]: ok, done

File: .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateExtractorConfiguration.cs [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+                DateTimeDefinitions.DateExtractor9,
+                RegexOptions.IgnoreCase | RegexOptions.Singleline),
+
+            // 2015-12-23
+            new Regex(
+                DateTimeDefinitions.DateExtractorA,
+                RegexOptions.IgnoreCase | RegexOptions.Singleline)
+        };
+
+
```
[the remark this answers is not in the document]
Comment (maurobennici) [reply .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateExtractorConfiguration.cs]: done

**Comment by tellarin** [comment]
Hi @maurobennici,

do you think you'll be able to cleanup the PR a little in the next coming days?

If we undo the regex deletions and add the "not supported" notes on the test Specs json files, we can already merge the initial structure.

Thanks.

**Comment by tellarin** [comment]
Hi @maurobennici,
I've fixed the merge conflicts and I'll try to merge as much of it as part of this PR or as a new PR, ok?
Thanks for the contribution! Do you think you'll be able to help continuing Italian support?

**Comment by tellarin** [comment]
Closing this one as PR #789 brings it up-to-date to merge.
